tests/stdlib/trstgen.nim

```diff
@@ -8,8 +8,37 @@ import ../../lib/packages/docutils/rstgen
 import ../../lib/packages/docutils/rst
 import unittest, strutils, strtabs
 
-proc toHtml(input: string): string =
-  rstToHtml(input, {roSupportMarkdown}, defaultConfig())
+proc toHtml(input: string,
+            rstOptions: RstParseOptions = {roSupportMarkdown},
+            expectError: string = "",
+            expectWarnings: seq[string] = @[]): string =
+  var warnings: seq[string]
+  var error: string
+  proc testMsgHandler(filename: string, line, col: int, msgkind: MsgKind,
+                      arg: string) =
+    let mc = msgkind.whichMsgClass
+    let a = $msgkind % arg
+    let message = "$1($2, $3) $4: $5" % [$filename, $line, $col, $mc, a]
+    if mc == mcError:
+      error = message
+      # we check only first error because subsequent ones may be meaningless
+      raise newException(EParseError, message)
+    else:
+      warnings.add message
+  try:
+    result = rstToHtml(input, rstOptions, defaultConfig(), testMsgHandler)
+  except EParseError:
+    discard
+  template assertEqual(s1, s2: untyped) =
```

Comment:
use unittest.check or unittest.require, don't roll your own (and assertEqual has the multiple template evaluation bug)

Reply:
Well, I switched to `check`, OK.

However for code
```
    check(warnings8[] == @["input(4, 0) Warning: unknown substitution " &
            "\'citation-som\'h"])
```
it seems very verbose:
```
    /home/amakarov/tmp/Nim/tests/stdlib/trstgen.nim(941, 22): Check failed: warnings8[] ==
    @["input(4, 0) Warning: unknown substitution " & "\'citation-som\'h"]
    warnings8[] was @["input(4, 0) Warning: unknown substitution \'citation-som\'"]
    @["input(4, 0) Warning: unknown substitution " & "\'citation-som\'h"] was @["input(4, 0) Warning: unknown substitution \'citation-som\'h"]
```

The constant in `@[...]` is repeated 3 times. Do you think `check` had better be improved?

Reply:
there are 2 things going on here:
* you're using `&` which prevent's check's eliding of `was` msg on lhs
* check's eliding doesn't work for more complex literals like seq, see https://github.com/timotheecour/Nim/issues/650